pkg/ibm/ibmp.go

```diff
@@ -253,6 +261,45 @@ func (ibmp IBMPowerDynamicConfig) GetState(kubeClient client.Client, ctx context
 	return "OK", nil
 }
 
+// CleanUpVms deletes any VMs in the pw cloud instance that are associated with a non-existing Tekton TaskRun.
+// Each VM instance should have a tag with the associated TaskRun's namespace and name. This is compared to
+// existingTaskRuns, which is a map of namespaces to a list of TaskRuns in that namespace, to determine if this
+// VM's TaskRun still exists.
+func (pw IBMPowerDynamicConfig) CleanUpVms(ctx context.Context, kubeClient client.Client, existingTaskRuns map[string][]string) error {
+	// Get all VM instances
+	log := logr.FromContextOrDiscard(ctx)
+	log.Info("Attempting to clean up orphaned IBM Power Systems instances")
+	service, err := pw.createAuthenticatedBaseService(ctx, kubeClient)
+	if err != nil {
+		return fmt.Errorf("failed to create an authenticated base service: %w", err)
+	}
+	pvmInstancesCollection, err := pw.listInstances(ctx, service)
+	if err != nil {
+		return fmt.Errorf("failed to fetch Power Systems instances: %w", err)
+	}
+
+	// Iterate over all VM instances
+	for _, instanceRef := range pvmInstancesCollection.PvmInstances {
+		pvmInstance, err := pw.getInstance(ctx, service, *instanceRef.PvmInstanceID)
+		if err != nil {
+			msg := fmt.Sprintf("failed to get instance %s; skipping this instance...", *instanceRef.PvmInstanceID)
+			log.Error(err, msg)
+			continue
+		}
+		// Delete the VM instance if the TaskRun namespace or TaskRun does not exist
+		taskRunExists := pw.doesInstanceHaveTaskRun(pvmInstance, log, existingTaskRuns)
+		if !taskRunExists {
+			err := pw.TerminateInstance(kubeClient, ctx, cloud.InstanceIdentifier(*pvmInstance.PvmInstanceID))
+			if err != nil {
+				log.Error(err, "failed to terminate instance", "instanceID", *pvmInstance.PvmInstanceID)
+				return fmt.Errorf("failed to terminate instance %s: %w", *pvmInstance.PvmInstanceID, err)
```

Comment:
Same here: should we aggregate errors and attempt to terminate all known instances?